### PR TITLE
KAFKA-5829; Speedup broker startup after unclean shutdown by reducing unnecessary snapshot files deletion

### PR DESCRIPTION
**Pull request by lindong28**

### CODE REVIEW

**Comment by lindong28** [comment]
The current Kafka implementation will cause slow startup after unclean shutdown. The time to load a partition will be 10X or more than what it actually needs. Here is the explanation with example:

- Say we have a partition of 20 segments, each segment has 250 message starting with offset 0. And each message has 1 MB bytes.
- Broker experiences hard kill and the index file of the first segment is corrupted.
- When broker startup and load the first segment, it realizes that the index of the first segment is corrupted. So it calls `log.recoverSegment(...)` to recover this segment. This method will call `stateManager.truncateAndReload(...)` which deletes the snapshot files whose offset is larger than base offset of the first segment. Thus all snapshot files are deleted.
- To rebuild the snapshot files, the `log.loadSegmentFiles(...)` will have to read every message in every segment of this partition even if their log and index files are not corrupted. This will increase the time to load this partition by more than an order of magnitude.

Note that there is similar performance issue if there is unclean broker shutdown and all existing producer state snapshot files' offsets are larger than recovery offset of the partition.

In order to address this issue, one simple solution is not to delete snapshot files that are than the given offset if only the index files needs re-build. More specifically, we should not need to re-build producer state snapshot file unless the related log file is corrupted or truncated.

**Comment by lindong28** [comment]
@ijuma @junrao Do you have time to review this patch?

**Comment by ijuma** [comment]
@hachikuji  should probably review this one.

**Comment by hachikuji** [comment]
@lindong28 Thanks for the patch. High level the approach sounds reasonable. Out of curiosity, did you actually hit this? I would have expected old segment index corruption to be rare in practice since the data is not changing. If this is a common scenario, then it may be useful to have this patch in 0.11.0.1.

**Comment by lindong28** [comment]
@hachikuji Yeah I am actually hitting this while testing kafka 0.11 for use at LinkedIn. It took me roughly 1 day to track down the issue and figure out the right solution. Note that I am testing Kafka with a hotfix patch as you can find in KAFKA-5297. I am not sure if this hotfix patch has increased the chance of hitting this issue.

**Comment by lindong28** [comment]
@hachikuji Note that this issue can also happen if there is unclean broker shutdown and all existing producer state snapshot files' offsets are larger than recovery offset of the partition. It doesn't necessarily require you to have old segment index corruption to hit this issue.

File: core/src/main/scala/kafka/log/ProducerStateManager.scala [hunk 1]
```diff
@@ -661,8 +661,8 @@ class ProducerStateManager(val topicPartition: TopicPartition,
       None
   }
 
-  private def latestSnapshotFile: Option[File] = {
-    val files = listSnapshotFiles
+  private def latestSnapshotFile(maxSnapshotOffset: Long): Option[File] = {
+    val files = listSnapshotFiles.filter(file => offsetFromFilename(file.getName) <= maxSnapshotOffset)
     if (files.nonEmpty)
       Some(files.maxBy(file => offsetFromFilename(file.getName)))
     else
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/ProducerStateManager.scala:703]: I removed my suggestion as it was not quite right. It's a bit annoying that we are calling `offsetFromFilename` on both iterations. It may make sense to change `listSnapshotFiles` to return a `Map[File, Long]` or `Seq[File, Long]]`.

**Comment by hachikuji** [comment]
@lindong28 Hmm... I might be wrong, but I think the unclean broker shutdown case should be rare too. We attempt to keep snapshots corresponding to the start of the last two segments, but we don't actually remove the snapshots for old segments until the logs have been flushed to disk. The intention was to always have a snapshot corresponding to the start of the segment containing the recovery point. Is this the case that you have been hitting? It would be helpful to understand a little bit better where this logic is failing. If you can reproduce the issue, maybe we can see which snapshots remain after unclean shutdown and what the recovery point is?

File: core/src/main/scala/kafka/log/ProducerStateManager.scala (hunk 1 above)
Comment (lindong28) [reply core/src/main/scala/kafka/log/ProducerStateManager.scala:703]: @ijuma Thanks for the comment. I will look into this.

**Comment by lindong28** [comment]
@hachikuji In the release certification test I manually killed the broker and see how much time it takes to startup. That may address your comment regarding the probability of unclean broker shutdown. I think you can probably reproduce the issue similarly with hard kill -- if not then I will try to reproduce it and see if it is more likely to happen due to some internal hot-fixes at LinkedIn. I will do that later after I finish the current work.

Does this sound OK? Do you prefer to commit this patch now or wait until I reproduce it?

**Comment by hachikuji** [comment]
@lindong28 If you don't mind, I'd prefer to wait until we have some more detail on the case where this is happening. It doesn't sound like a blocker for 0.11.0.1 from what I've heard so far, but I'm a little concerned that there may be a bug in a snapshot retention logic. I will attempt to reproduce the issue here as well.

**Comment by lindong28** [comment]
@hachikuji OK. I will ask you for more detail about the information you need when I am ready to reproduce it.

**Comment by ijuma** [comment]
It would be good to include this in 1.0.0. @lindong28, do you still have access to the environment where you saw this?

**Comment by lindong28** [comment]
@ijuma No.. I am not able to remotely access dev test cluster from Shanghai. Previously I was too busy with release certification and KIP-113 implementation and therefore didn't have time to investigate this. While it may be useful to get more information, I think it may be OK to merge this patch in its current state, given that Jason and I both agree this is a problem based on the code. If the test suggested by Jason is required, I will only be able to do it after 1.0 release.

**Comment by ijuma** [comment]
@hachikuji, thoughts?

**Comment by hachikuji** [comment]
@ijuma I will try to reproduce the issue today. My concern is that we don't know specifically where the current logic is incomplete (and I'm not sure whether LI hotfixes were ever ruled out as the cause).

**Comment by hachikuji** [comment]
Ok, I think I get it now. The issue is a result of the fact that the recovery checkout file will generally be out of sync with the actual flushed offset. By default, the interval that we write the checkpoint file is one minute (seems a bit low btw), so it could be pretty far behind when the broker is hard-killed. This probably means we're being a bit too greedy deleting snapshot files immediately when the flush point advances; it might be better to wait until the checkpoint has been written to disk.

The other issue which this patch is attempting to address is avoiding the removal of snapshot files unnecessarily when only the index files change. At a high level, this seems like a good idea, but I'm not sure that the current patch achieves it. Although it prevents the deletion of the latest snapshot files, it does not do anything to prevent the reloading itself (which is the really expensive part).

In fact, if you think about it, rebuilding the producer state is not that big of a deal if we are going to the trouble of validating the segments after the recovery offset anyway. The main issue is that rebuilding the producer state requires us to load all segments _before_ the recovery offset. So I think the proper fix is what I suggested above: avoid deletion of snapshot files until the recovery point has actually been written to disk.

**Comment by lindong28** [comment]
@hachikuji Thanks for the detailed explanation.

Your answer explained why the current Kafka implementation will delete all available snapshot files after a hard kill. For ease of understanding, I will summarize the sequence of events that will cause the problem:

1) Broker flushes data after offset 1000 to disk and removed snapshot files before offset 1000. It also increases the high watermark in memory but it has not been written to disk yet. The high watermark in disk is still 500.
2) Broker is hard killed. 
3) Broker restarts. Now it will attempt to recovery segments after offset 500, which is the high watermark in the disk. Current Kafka implementation will delete snapshot files with offset after the recovery offset. Thus all snapshot files are deleted.
4) Then broker will need to read ALL segments of this partition to rebuild snapshot files. This is the expensive part and it is what we need to fix.

The solution in this patch is to avoid deleting snapshot file in segment recovery, unless the log segment is found to be corrupted during segment recovery. This patch will address the problem by not deleting snapshot file in the step 3) above. Log segment before offset 1000 should not be corrupted after the data is flushed to disk. Can you explain a bit more why you are not sure the current patch achieves it?

I think your solution will also work. It addresses the problem by not deleting snapshot files before offset 1000 in step 1) until recovery offset has been written. However, it may be a bit unintuitive to couple the snapshot file management with recovery offset management. Ideally snapshot file should depend only on the content of the log segment. If all data in the log segment has been flushed to disk, we should be able to safely delete old snapshot files.

The current approach in the patch choose not to delete snapshot files unless the log segment is corrupted, which makes sense, right? And if it actually addresses the problem, I am not sure we need to have snapshot file management depend on the recovery offset checkpoint interval. What do you think?

**Comment by hachikuji** [comment]
@lindong28 When we recover a log, we take snapshots after every segment has been recovered. By the time we have recovered all log segments, the snapshots are up-to-date. That means that once recovery has finished, there is virtually no cost to loading producer state (other than what we payed during recovery). It does not matter whether we skip deletion of snapshots greater than the recovery point because we will rebuild them anyway.

The point is that once we have decided we need to do recovery, then we have to scan each log segment greater than the recovery offset in order to validate the log and rebuild the index files. It is the scanning of the log segments that causes the slow startup. Your patch prevents the deletion of snapshots greater than the recovery point during recovery, but if we're already scanning all the log segments anyway (because they are beyond the recovery point), then what's the point? I checked out your patch locally and hard-killed the broker. The behavior is identical to what I saw without the patch.  

Rebuilding producer state if we are already scanning the log is not a major concern. The one reason why there is an issue here as far as I can tell is the fact that we will have to scan the segments from before the recovery offset as well in order to rebuild producer state. Again, this patch does not solve that problem. The only fix I can see is to prevent the deletion of snapshots until we know the recovery offset has been written to disk.

**Comment by hachikuji** [comment]
cc @junrao (since you reviewed the patch for producer snapshot recovery)

**Comment by lindong28** [comment]
@hachikuji 

"Your patch prevents the deletion of snapshots greater than the recovery point during recovery, but if we're already scanning all the log segments anyway (because they are beyond the recovery point), then what's the point"

I think the point of not deleting snapshots greater than the recovery point is so that broker does not need to scan all log segments before the recovery point. Does this make sense? Can you help me understand if I missed anything here?

"The one reason why there is an issue here as far as I can tell is the fact that we will have to scan the segments from before the recovery offset as well in order to rebuild producer state. Again, this patch does not solve that problem."

Previously I explained the sequence of events to show why it is currently a problem, and explained why this patch could fix this problem. Could you please be more specific why this patch does not solve that problem? Thanks.

**Comment by hachikuji** [comment]
> I think the point of not deleting snapshots greater than the recovery point is so that broker does not need to scan all log segments before the recovery point. Does this make sense? Can you help me understand if I missed anything here?

Look at the code. When we recover a segment, we have to first load the producer state up to the base offset of that segment. That means scanning all the segments which come before it. It does not matter what snapshots we retain after the recovery point. Even with your patch, we still have to do this. 

Say we have segments 0, 10, 20, 30, and 40. Suppose that we have snapshots for segments 30 and 40, but when the broker is hard-killed, the recovery offset is only at 10. When we recover the log, we will start by recovering segment 10. To do so, we will first load producer state from the beginning of the log, which means we have to scan from segment 0. Once we have recovered segment 10, we will take a snapshot of the producer state at offset 20. We will then go on to recover segment 20. This time loading producer state will be easy because we took a snapshot at offset 20. Once we load producer state from segment 20, we'll take a new snapshot at offset 30. This goes on until we have recovered all segments. 

The main point is that recovering the first segment after the recovery point (10 in this case) requires us to scan all prior segments. Your patch does not change this.

**Comment by lindong28** [comment]
@hachikuji Now I understand the difference between our understanding of the fix.

The current patch will avoid slow start in your example because when Log.recoverSegment(...) is called for the segment 10, this method will not delete snapshots with offset 30 and 40. Thus stateManager.mapEndOffset() will be 40, `logSegments(stateManager.mapEndOffset, segment.baseOffset)` will be empty, and segments before offset 10 will not need to be scanned.

When we recover a segment, do you think it makes sense to skip re-building snapshot for this segment as long as there exists snapshot for a later segment?

**Comment by hachikuji** [comment]
> The current patch will avoid slow start in your example because when Log.recoverSegment(...) is called for the segment 10, this method will not delete snapshots with offset 30 and 40. Thus stateManager.mapEndOffset() will be 40, logSegments(stateManager.mapEndOffset, segment.baseOffset) will be empty, and segments before offset 10 will not need to be scanned.

Let me try to explain again. The loading of producer state happens during recovery, not after. When we recover the first segment that needs it, we check for snapshot files that we can use. In the hard failure case, we may not have any, so we must load producer state from the first segment: https://github.com/apache/kafka/pull/3782/files#diff-ffa8861e850121997a534ebdde2929c6R343. So during recovery of segment 10, we load segment 0. It does not matter if we retain segment 30 and 40 after recovering segment 10 because the damage is already done. That is why your patch does not solve the problem.

>When we recover a segment, do you think it makes sense to skip re-building snapshot for this segment as long as there exists snapshot for a later segment?

I thought about this, but the problem is we don't know whether there will be corrupt data which will invalidate the later snapshots. In the worst case, we'll have to re-read the segments. As I said before, as long as we are already reading through a segment in order to rebuild the index files, reloading the producer state as well is a small cost

I discussed this offline with @junrao. Another possibility is that we just write the recovery checkpoint file every time we flush the logs in `LogManager`. In general, the recovery point should be fairly small even if you have a few thousands of partitions, so maybe the cost is not a major concern. This would reduce the window for this edge case.

**Comment by lindong28** [comment]
"In the hard failure case, we may not have any".

@hachikuji Why we will not have any snapshot files in case of the hard failure? Doesn't Kafka broker always keep at least 1 snapshot file regardless of how it flushes data to disk or checkpoints recovery offset?

**Comment by hachikuji** [comment]
@lindong28 There are two rules in the current logic:

1. We always keep snapshots from the latest two segments.
2. We do not delete old snapshots until the flush point has advanced beyond them.

The problem is that the flush point can advance before the recovery offsets have been written to the `recovery-point-offset-checkpoint` file. There will generally be a delay because writing of this file happens asynchronously. In the case of a hard-kill, it is very likely that a partition's flush point has gotten well ahead of the offset in `recovery-point-offset-checkpoint`. So there are basically two solutions:

1. We do not delete old snapshots until the flush point has been written to `recovery-point-offset-checkpoint`. 
2. We always write `recovery-point-offset-checkpoint` soon after flushing. There is still a window here where a hard failure could result in unnecessary loading, but we have reduced it considerably.

**Comment by lindong28** [comment]
@hachikuji I think I understand every single bit of your summary. And I still think my patch addresses the problem given that broker always keeps snapshots from the latest two segments. My point is that, regardless of when the recovery offset is written, during segment recovery, if there exists a snapshot whose offset is large than this segment (which exists because the latest two snapshots are retained), there will be no need to rebuild snapshot for this segment, and therefore is no need to re-read data since the first segment. Can you tell me what's wrong with this reasoning?

**Comment by hachikuji** [comment]
@lindong28 Please look at the code:

1. https://github.com/apache/kafka/pull/3782/files#diff-ffa8861e850121997a534ebdde2929c6R342: On this line, we load producer state up to the segment that is being recovered. In the case we're talking about, the segment that is being loaded is prior to any of the existing snapshots, so when we initialize producer state here, it will be empty and the last "map offset" will be equal to the log start offset.

2. https://github.com/apache/kafka/pull/3782/files#diff-ffa8861e850121997a534ebdde2929c6R343: Now we load producer state from the last offset in the producer state, which is the log start offset since we had no snapshot to load from. So we scan all segments prior to the current segment in order to build the producer state prior to the segment being recovered.

3. https://github.com/apache/kafka/pull/3782/files#diff-ffa8861e850121997a534ebdde2929c6R354: Finally, after we have loaded the segment, we take a snapshot. Now when we go on to recover the next segment, we will have this snapshot to start from.

When all of this finishes, we will have snapshots corresponding to the base offset of every segment that we recovered. Because we have to load the state from the beginning offset in step 2, it does not matter whether we retain the snapshots after the recovery point. The snapshot files only help for recovering data at offsets larger than the snapshot offset.

**Comment by lindong28** [comment]
@hachikuji I actually read the code and understand every bit you are saying. I think you are wrong in the step 2 of your explanation. In step 2 and line https://github.com/apache/kafka/pull/3782/files#diff-ffa8861e850121997a534ebdde2929c6R343, stateManager.mapEndOffset will be the offset of the latest snapshot, which should be the offset of the latest segment given that broker will maintain the latest two snapshots. Thus `logSegments(stateManager.mapEndOffset, segment.baseOffset)` will be empty, and broker does not need to scan all segments prior to the current segment.

Note that I have explicitly mentioned this in my previous reply. You can find it by searching for "Thus stateManager.mapEndOffset() will be 40" in this github webpage.

Also, I think we should find the right solution in the design level rather than sticking to the current implementation. The idea is that, if there exists snapshot with offset larger than the current segment, there is no need to re-build snapshot for the current segment, thus no need to scan segments prior to the current segment. If the current implementation does not do it, we can fix it so that it does what we want.

BTW, I am pretty sure the problem goes away after I apply this patch based on my prior experiment.

**Comment by hachikuji** [comment]
Let me give you a simple reproducible test. I took your existing patch and added the following println:

```scala
    val stateManager = new ProducerStateManager(topicPartition, dir, maxProducerIdExpirationMs)
    stateManager.truncateAndReload(logStartOffset, segment.baseOffset, time.milliseconds, deleteFilesAboveMaxSnapshotOffset = false)
    println(s"Map end offset is ${stateManager.mapEndOffset} before recovering segment ${segment.baseOffset}")
```

I started a producer and let it roll the log a few times. Then I hard-killed the broker. Here are the segments in the log after the hard kill:
```
-rw-r--r--   1 jason  wheel      552184 Oct  2 18:27 00000000000000000000.index
-rw-r--r--   1 jason  wheel  1073737344 Oct  2 18:27 00000000000000000000.log
-rw-r--r--   1 jason  wheel      140760 Oct  2 18:27 00000000000000000000.timeindex
-rw-r--r--   1 jason  wheel      552184 Oct  2 18:27 00000000000001035360.index
-rw-r--r--   1 jason  wheel  1073737344 Oct  2 18:27 00000000000001035360.log
-rw-r--r--   1 jason  wheel       80016 Oct  2 18:27 00000000000001035360.timeindex
-rw-r--r--   1 jason  wheel      552184 Oct  2 18:27 00000000000002070720.index
-rw-r--r--   1 jason  wheel  1073737344 Oct  2 18:27 00000000000002070720.log
-rw-r--r--   1 jason  wheel       78804 Oct  2 18:27 00000000000002070720.timeindex
-rw-r--r--   1 jason  wheel      552184 Oct  2 18:27 00000000000003106080.index
-rw-r--r--   1 jason  wheel  1073732179 Oct  2 18:27 00000000000003106080.log
-rw-r--r--   1 jason  wheel       92388 Oct  2 18:27 00000000000003106080.timeindex
-rw-r--r--   1 jason  wheel      552184 Oct  2 18:27 00000000000004141435.index
-rw-r--r--   1 jason  wheel  1073737344 Oct  2 18:27 00000000000004141435.log
-rw-r--r--   1 jason  wheel          10 Oct  2 18:27 00000000000004141435.snapshot
-rw-r--r--   1 jason  wheel       84108 Oct  2 18:27 00000000000004141435.timeindex
-rw-r--r--   1 jason  wheel    10485760 Oct  2 18:27 00000000000005176795.index
-rw-r--r--   1 jason  wheel   536993120 Oct  2 18:27 00000000000005176795.log
-rw-r--r--   1 jason  wheel          10 Oct  2 18:27 00000000000005176795.snapshot
-rw-r--r--   1 jason  wheel    10485756 Oct  2 18:27 00000000000005176795.timeindex
-rw-r--r--   1 jason  wheel           8 Oct  2 18:26 leader-epoch-checkpoint
```
And here is the content of the recovery-checkpoint file:
```
0
1
foo 0 2070720
```
Note that the checkpointed offset 2070720 is considerably lower than the last snapshot offset 4141435. I then restarted the broker. Here is the output:

```
[2017-10-02 18:30:48,294] INFO Recovering unflushed segment 2070720 in log foo-0. (kafka.log.Log)
Map end offset is 0 before recovering segment 2070720
[2017-10-02 18:31:12,534] INFO Recovering unflushed segment 3106080 in log foo-0. (kafka.log.Log)
[2017-10-02 18:31:12,535] INFO Loading producer state from snapshot file '/tmp/kafka-logs/foo-0/00000000000003106080.snapshot' for partition foo-0 (kafka.log.
ProducerStateManager)
Map end offset is 3106080 before recovering segment 3106080
[2017-10-02 18:31:15,435] INFO Recovering unflushed segment 4141435 in log foo-0. (kafka.log.Log)
[2017-10-02 18:31:15,436] INFO Loading producer state from snapshot file '/tmp/kafka-logs/foo-0/00000000000004141435.snapshot' for partition foo-0 (kafka.log.ProducerStateManager)
Map end offset is 4141435 before recovering segment 4141435
[2017-10-02 18:31:18,183] INFO Recovering unflushed segment 5176795 in log foo-0. (kafka.log.Log)
[2017-10-02 18:31:18,184] INFO Loading producer state from snapshot file '/tmp/kafka-logs/foo-0/00000000000005176795.snapshot' for partition foo-0 (kafka.log.
ProducerStateManager)
Map end offset is 5176795 before recovering segment 5176795
[2017-10-02 18:31:18,855] INFO Loading producer state from offset 5694595 for partition foo-0 with message format version 2 (kafka.log.Log)
[2017-10-02 18:31:18,857] INFO Loading producer state from snapshot file '/tmp/kafka-logs/foo-0/00000000000005694595.snapshot' for partition foo-0 (kafka.log.
ProducerStateManager)
[2017-10-02 18:31:18,858] INFO Completed load of log foo-0 with 6 log segments, log start offset 0 and log end offset 5694595 in 31676 ms (kafka.log.Log)
```

This is working exactly as it is supposed to. When recovering each segment, we load producer state up to the base offset of that segment.

**Comment by lindong28** [comment]
@hachikuji I am not sure we are on the same page regarding what we are discussing. The bug we have now in Kafka is that, broker will have to scan all log segments of a partition, including those before the recovery offset, upon hard kill. This patch fixes this issue by saying that only segments after the recovery offset needs to be scanned. Therefore, if you want to show my patch does not work, you probably want to show that segment `00000000000000000000` and `00000000000001035360` is scanned. I do not see it in your log.

Honestly speaking, I think you have focused more on showing that this patch does not work than trying to find the right fix for this issue. I have been trying to explain to you that the right fix is to skip re-building snapshot for a segment if there exists snapshot with later offset. If you think this is the right solution, all we need is to make sure that the code does it. I don't understand why we need a test to show that the code does not work instead of first agreeing on the right solution.

@junrao @ijuma Do you mind chime in, take a look at this patch and the discussion, and comment on the right solution? We are getting close to the deadline of code freeze. My availability may be limited in the following days. So it is better to agree on the solution earlier than later. Thanks!
